---
title: "R_Assignment"
author: "ABDULHAKIM"
date: "3-3-2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
setwd('.')

```

## Reading the Genotypes and SNPs file
here we load the files fang_et_al_genotypes.txt and snp_position file 
we assign the files as variables: fang_genotypes and SNPs
```{r}
fang_genotypes <- read.table("https://raw.githubusercontent.com/EEOB-BioData/BCB546_Spring2024/main/assignments/UNIX_Assignment/fang_et_al_genotypes.txt", header = TRUE, sep = "\t")
SNPs <- read.table("https://raw.githubusercontent.com/EEOB-BioData/BCB546_Spring2024/main/assignments/UNIX_Assignment/snp_position.txt", header = TRUE, sep = "\t")
```

# Part 1
## DATA INSPECTION

We were able to analyze both the genotypes and SNP files and the code for that is below. 

```{r}

dim(fang_genotypes)
#we find dimensions of the fang_genotypes file which is 2782 rows and 986 columns
dim(SNPs)
#from this we get the dimensions of the SNPs file: 983 rows and 15 columns

fang_genotypes[1:10, 1:10]
SNPs[1:10, 1:10]
#here we extract a subset of the first 10 rows and the first 10 columns and helps us get an idea of what the file looks like from the entire table. 

class(fang_genotypes)
class(SNPs)
#from this we find information about the vectors and we can see that both files are data.frame format. 

View(fang_genotypes)
View(SNPs)
#from this function we can inspect the both files in a separate tab

str(fang_genotypes)
str(SNPs)
#we can also use this function to get a summary information about structure of the data frame such as name, type and preview of data in each column. 

```
## PART 2
## Data Processing

Sort out the genotypes based on the Group and create separate data files of each species. 

```{r}

library(dplyr)

colnames(fang_genotypes)
#from this we can find all the column names that exist in fang_genotypes

maize_group <- filter(fang_genotypes, Group == 'ZMMIL' | Group == 'ZMMLR' | Group == 'ZMMMR')
#here we filter out the maize data and assign variable names

teosinte_group <- filter(fang_genotypes, Group == 'ZMPBA' | Group == 'ZMPIL' | Group == 'ZMPJA')
#we do same for the Teosinte Data and assign variable name

str(maize_group)
str(teosinte_group)
#by using this function we can double check which groups we have filtered and we have successfully filtered out the only groups we need
```

### Transposing the Maize Data

```{r}
install.packages("tibble")  # Install tibble package 
library(tibble)  # Load the tibble package 

maize_group <- column_to_rownames(maize_group, var = "Sample_ID")
#here we reassign the sample id names from column to rows

transposed_maize <- t(maize_group) %>% as.data.frame() %>% rownames_to_column(., var = "SNP_ID") 
#we assigned a new variable transposed_maize, for the new transposed maize group then convert the transposed object to a data frame
view(transposed_maize)
#here we can view the final transposed_maize we can see the first three are lines are titles

transposed_maize <- transposed_maize[3:nrow(transposed_maize),]
#here we remove the first three rows as to get it ready to merge the file

Maize_data <- merge(SNPs, transposed_maize, by = "SNP_ID")
#here we merge the files based on matching SNP_ID values

Maize_data <- select(Maize_data, SNP_ID, Chromosome, Position, everything())
#lastly we reorder the columns starting with SNP_ID,chromosome and position
View(Maize_data)
#we can double check the result here
```

### Tranposing the Teosinte Data
we do the same thing for the Teosinte group and merge the file

```{r}
teosinte_group <- column_to_rownames(teosinte_group, var = "Sample_ID")
transposed_teosinte <- t(teosinte_group) %>% as.data.frame() %>% rownames_to_column(., var = "SNP_ID") 
transposed_teosinte <- transposed_teosinte[3:nrow(transposed_teosinte),]
Teosinte_data <- merge(SNPs, transposed_teosinte, by = "SNP_ID")
Teosinte_data <- select(Teosinte_data, SNP_ID, Chromosome, Position, everything())

```

## Maize Output

```{r}
chr_maize <- filter(Maize_data, Chromosome != "multiple" & Chromosome != "unknown")
str(chr_maize)
#here the code filters `Maize_data` to exclude rows where the Chromosome is "multiple" or "unknown." then using the str function we can get a brief summary of what chr_maize looks like

dir.create("./Maize", recursive = TRUE, showWarnings = FALSE)
#we create a new directory called Maize

for (i in 1:length(unique(chr_maize$Chromosome))){
  chrm <-  chr_maize %>% filter(Chromosome == i) %>% arrange(Position)
  write.table(chrm, file = paste("./Maize/Maize_inc_A",i, sep="_"))
}

#Then here using for loop the sequence is generated for each unique chromosome,then  it filters `chr_maize` for rows matching that chromosome, sort them by "Position" column in increasing order then writes each chromosome's data to a file named "Maize_inc_A_[chromosome number]" in the "Maize" directory.

```

```{r}
Subbed_maize <- as_tibble(lapply(Maize_data, gsub, pattern ="?", replacement ="-", fixed = TRUE))
#the new 'Subbed_maize' is a new data where we are substituting the '?' with '-'  converting the result into a tibble

Subbed_Chr_Maize <- filter(Subbed_maize, Chromosome != "multiple" & Chromosome != "unknown")
str(Subbed_Chr_Maize)
#here we filter out the chromosome with multiple or unknown

for (i in 1:length(unique(Subbed_Chr_Maize$Chromosome))) {
  chrm_subbed <- Subbed_Chr_Maize %>% filter(Chromosome == i) %>% arrange(desc(Position))
  write.table(chrm_subbed, file = paste("./Maize/Maize_dec_B", i, sep = "_"))
  }
#here we do the same using the for loop as the previous part but with decreasing position
```

## Teosinte Output

```{r}
chr_teosinte <- filter(Teosinte_data, Chromosome != "multiple" & Chromosome != "unknown")
str(chr_teosinte)
#here the code filters `Teosinte_data` to exclude rows where the Chromosome is "multiple" or "unknown." then using the str function we can get a brief summary of what chr_maize looks like

dir.create("./Teosinte", recursive = TRUE, showWarnings = FALSE)
#we create a new directory called Teosinte

for (i in 1:length(unique(chr_teosinte$Chromosome))) {
  chrt <- chr_teosinte %>% filter(Chromosome == i) %>% arrange(Position)
  write.table (chrt, file = paste("./Teosinte/Teosinte_inc_A", i, sep = "_"))
}
#Then here using for loop the sequence is generated for each unique chromosome,then  it filters `chr_teosinte` for rows matching that chromosome, sort them by "Position" column in increasing order then writes each chromosome's data to a file named "Teosinte_inc_A_[chromosome number]" in the "Teosinte" directory
```

```{r}
Subbed_teosinte <- as_tibble(lapply(Teosinte_data, gsub, pattern = "?", replacement = "-", fixed = TRUE))
#the new 'Subbed_teosinte' is a new data same as the maize  we are substituting the '?' with '-'  converting the result into a tibble

Subbed_Chr_Teosinte <- filter(Subbed_teosinte, Chromosome != "multiple" & Chromosome != "unknown")
#here we filter out the chromosome with multiple or unknown

for (i in 1:length(unique(Subbed_Chr_Teosinte$Chromosome))) {
    chrt_subbed <- Subbed_Chr_Teosinte %>% filter(Chromosome == i) %>% arrange(desc(Position))
    write.table(chrt_subbed, file = paste("./Teosinte/Teosinte_dec_B", i, sep = "_"))
}
#here we do the same using the for loop as the previous part but with decreasing position for teosinte
```

# PART 3
# DATA VISUALIZATION
clean up and formating SNPs and Fang genotype data frames
```{r}
SNP_clean <- SNPs %>% select(SNP_ID, Chromosome, Position)
#we create a new clean data frame that includes only the SNP_ID, chromosome and position

transposed_fang_genotypes <- fang_genotypes %>% select(-JG_OTU, -Group) %>% column_to_rownames(., var = "Sample_ID") %>% t() %>% as.data.frame() %>% rownames_to_column(., var = "SNP_ID")
#here to clean up we exclude JG_OTU and the Group from fang_genotypes data frame then we transpose the new data 'transposed_fang_genotypes' so now it will only have the SNP_ID and other except for JG_oTU and Group
view(transposed_fang_genotypes)

Merged_fang_genotypes <- merge(SNP_clean, transposed_fang_genotypes) %>% filter(., Chromosome != "unknown" & Chromosome != "multiple")
#here we merge the cleaned and transposed SNPs and fang_genotypes and filter out chromosomes that are unknown and multiple

```

SNPs per chromosome 

```{r}
library(ggplot2)

total_SNPs <- ggplot(Merged_fang_genotypes, aes(x= Chromosome)) + geom_bar(aes(fill = Chromosome)) + theme_bw() + labs(x = "Chromosome", y = "Total number of SNPs")
#here we create a bar plot to visualize the total SNPs for each chromosome in 'Merged_fang_genotypes' where x will be the chromosome and y is the total number of snps

view(Merged_fang_genotypes)
print(total_SNPs)
ggsave("total_SNPs.png",total_SNPs)

#finally we can print the bar plot for the total SNPs and save it to the directory


# Create a new dataframe with only numeric Position values from the merged_fang genotypes
numeric_fang_genotypes <- Merged_fang_genotypes %>%  mutate(Position_numeric = as.numeric(as.character(Position))) %>% filter(!is.na(Position_numeric)) %>%  select(SNP_ID, Chromosome, Position_numeric) %>%  rename(Position = Position_numeric)

view(numeric_fang_genotypes)

# Now we can plot using the new dataframe

Diversity_SNPs <- ggplot(numeric_fang_genotypes, aes(x = Position)) + 
  geom_density(aes(fill = Chromosome)) + 
  facet_wrap(~ Chromosome, nrow = 5, ncol = 2) + 
  theme_bw() + 
  labs(x = "Position", y = "Density")
#here i created a density plot to visualize the distribution of SNPs for each chromosome in Merged_fang_genotypes data

Diversity_SNPs


pdf("SNPs_per chromosome.pdf")
print(total_SNPs)
print(Diversity_SNPs)
ggsave("Diversity_SNPs.png",Diversity_SNPs)
dev.off()

```

## Visualization of missing data and amount of heterozygosity

```{r}

# Install tidyr 
install.packages("tidyr")

# Load tidyr
library(tidyr)

tidy_fang_genotypes <- fang_genotypes %>% select(-JG_OTU) %>% pivot_longer(cols=c(-Sample_ID, -Group), names_to = "SNP_ID", values_to = "Sequence")
#we create a new data frame 'tidy_fang_genotype' from the fang_genotypes by excluding JG_OTU column and pivot it to a long format where each row represents a unique combination of sample_ID,Group and SNP_ID with the corresponding sequence

tidy_fang_genotypes <- tidy_fang_genotypes %>% mutate(new_sequence = ifelse(Sequence %in% c("A/A", "T/T", "C/C", "G/G"), "Homozygous", ifelse(Sequence == "?/?", "Missing", "Heterozygous")))
#here we create new columns for sequences with same alleles 'Homozygous', 'Missing' for those with '?/?', and 'Heterozygous' for those with different alleles by using 'ifelse' and stores it a new column called new_sequence

View(tidy_fang_genotypes)
#overview of the created column classifying the genotypes

#For all the samples.
Samples_Plot <-  ggplot(tidy_fang_genotypes, aes(x = Sample_ID, fill = new_sequence)) + geom_bar(position = "fill") + theme_bw() + labs(x = "Sample ID", y = "Proportion")
#here we create a bar plot to show proportions of sequence for each Sample_ID in the tidy_fang_genotypes

#Stacked Bar-graph for all groups. 
Groups_Plot <- ggplot(tidy_fang_genotypes, aes(x = Group , fill = new_sequence)) + geom_bar(position = "fill") + 
  theme_bw() + theme(axis.text.x = element_text(angle = 90))+ labs(y = "Proportion")
#here we generate a bar plot to show proportions of the new_sequence column

pdf("Missing_DATA_Heterozygosity_Visualisation.pdf")
print(Samples_Plot)
print(Groups_Plot)
ggsave("Sample_Plot.png", Samples_Plot)
ggsave("Group_Plot.png",Groups_Plot)
dev.off()

```

##  My visualization for the proportion of Alleles in different species

```{r}
My_visualization <- ggplot(filter(tidy_fang_genotypes, Sequence != "?/?") , aes(x = Sample_ID, fill = Sequence)) + 
  geom_bar(position = "fill") + theme_bw() + labs(x = "Sample ID", y = "Proportion")
#here i compare the proportion of each alleles that exist in each chromosome except for the missing sequences

My_visualization

pdf("My_visualization.pdf")
ggsave("My_visualization.png", My_visualization)
dev.off()

```
